Plugins/NativeCamera/Source/Android/NativeCameraImpl.cpp

```diff
@@ -309,129 +256,248 @@ namespace Babylon::Plugins
+
+            // Get all available stream configurations supported by the camera
+            API24::ACameraMetadata_const_entry streamConfigurations = {};
+            GET_CAMERA_FUNCTION(ACameraMetadata_getConstEntry)(metadataObj, API24::ACAMERA_SCALER_AVAILABLE_STREAM_CONFIGURATIONS, &streamConfigurations);
+            // format of the data:
+            // 0: format
+            // 1: width
+            // 2: height
+            // 3: input?
+
+            for (uint32_t j = 0; j < streamConfigurations.count; j += 4) {
+                int32_t format{streamConfigurations.data.i32[j + 0]};
+                int32_t width{streamConfigurations.data.i32[j + 1]};
+                int32_t height{streamConfigurations.data.i32[j + 2]};
+                int32_t input{streamConfigurations.data.i32[j + 3]};
+
+                if (input || format != AIMAGE_FORMAT_YUV_420_888)
+                {
+                    // Ignore the configuration if it is either an input type or not a preview format
+                    continue;
+                }
+
+                auto cameraResolution{ std::make_shared<CameraTrack>() };
+                cameraResolution->width = width;
+                cameraResolution->height = height;
+
+                cameraDevice->supportedResolutions.push_back(cameraResolution);
+            }
+
+            // Get camera hardware info
+            API24::ACameraMetadata_const_entry metaDataEntry = {};
+            GET_CAMERA_FUNCTION(ACameraMetadata_getConstEntry)(metadataObj, API24::ACAMERA_LENS_FACING, &metaDataEntry);
+            auto facing{ static_cast<API24::acamera_metadata_enum_android_lens_facing_t>(metaDataEntry.data.u8[0]) };
+
+            GET_CAMERA_FUNCTION(ACameraMetadata_getConstEntry)(metadataObj, API24::ACAMERA_SENSOR_ORIENTATION, &metaDataEntry);
+            int32_t sensorOrientation{ metaDataEntry.data.i32[0] };
+
+            GET_CAMERA_FUNCTION(ACameraMetadata_getConstEntry)(metadataObj, API24::ACAMERA_FLASH_INFO_AVAILABLE, &metaDataEntry);
+            bool torchSupported = metaDataEntry.data.u8[0];
+
+            GET_CAMERA_FUNCTION(ACameraMetadata_getConstEntry)(metadataObj, API24::ACAMERA_CONTROL_ZOOM_RATIO, &metaDataEntry);
+            float zoomRatio{ metaDataEntry.data.f[0] };
+
+            GET_CAMERA_FUNCTION(ACameraMetadata_getConstEntry)(metadataObj, API24::ACAMERA_CONTROL_ZOOM_RATIO_RANGE, &metaDataEntry);
+            float minZoomRatio{ metaDataEntry.data.f[0] };
+            float maxZoomRatio{ metaDataEntry.data.f[1] };
+
+            // update the cameraDevice information
+            cameraDevice->implData = std::make_unique<CameraDevice::ImplData>();
+            cameraDevice->implData->cameraID = id;
+            cameraDevice->implData->sensorRotation = sensorOrientation;
+            cameraDevice->implData->facingUser = facing == API24::ACAMERA_LENS_FACING_FRONT;
+
+            cameraDevice->capabilities.push_back(std::make_shared<CameraCapabilityTemplate<std::string>>(
```

Comment:
Same for all these places, I don't think you want to use make_shared but rather emplace_back a value owned by each vector.

Reply:
I have patched things up now so that I use emplace_back and capture the value if needed from that call. For the most part I have gotten rid of all shared_ptrs with only a few exceptions of pre-existing code that relies on it (NativeCameraImpl for instance).